backend/src/Database.hs

```diff
@@ -111,20 +116,83 @@ toFilterExpression ::
+      Just (InequalityFilter EQ a) -> (entity ^. field) ==. (just . val $ a)
+    toListFilter ::
+      (PersistEntity e, PersistField f) =>
+      SqlExpr (Entity e) -> EntityField e f -> [f] -> SqlExpr (Value Bool)
+    toListFilter entity field = ((entity ^. field) `in_`) . valList
+    filterList :: [Maybe (SqlExpr (Value Bool))]
+    filterList =
+      [ playerFilter,
+        pairingFilter,
+        timestampFilter,
+        winnerFilter,
+        loserFilter,
+        turnsFilter,
+        victoryFilter,
+        leagueFilter,
+        tokensFilter,
+        dwarvenRings,
+        corruption,
+        mordor,
+        aragorn,
+        treebeard,
+        initialEyes,
+        interestRating,
+        hasLog
+      ]
+    playerFilter = liftA2 (||.) (toListFilter report GameReportWinnerId <$> spec.players) (toListFilter report GameReportLoserId <$> spec.players)
     pairingFilter =
       spec.pairing <&> \case
         (player1, Nothing) -> (report ^. GameReportWinnerId ==. val player1) ||. (report ^. GameReportLoserId ==. val player1)
         (player1, Just player2) ->
           ((report ^. GameReportWinnerId ==. val player1) ||. (report ^. GameReportLoserId ==. val player1))
             &&. ((report ^. GameReportWinnerId ==. val player2) ||. (report ^. GameReportLoserId ==. val player2))
-    winnerFilter = toFilter report GameReportWinnerId spec.winners
-    loserFilter = toFilter report GameReportLoserId spec.losers
-    leagueFilter = case spec.leagues of
-      Just [] -> Just (isNothing_ (report ^. GameReportLeague))
-      _ -> toFilter report GameReportLeague (map Just <$> spec.leagues)
+    timestampFilter =
+      spec.timestamp <&> \case
+        Before t -> report ^. GameReportTimestamp <=. val t
+        After t -> report ^. GameReportTimestamp >=. val t
+        Between start end -> (report ^. GameReportTimestamp >=. val start) &&. (report ^. GameReportTimestamp <=. val end)
+    winnerFilter = toListFilter report GameReportWinnerId <$> spec.winners
+    loserFilter = toListFilter report GameReportLoserId <$> spec.losers
+    turnsFilter = toInequalityFilter report GameReportTurns <$> spec.turns
+    victoryFilter =
+      spec.victory
+        <&> foldr1 (||.)
+        . fmap (\(s, v) -> (report ^. GameReportSide) ==. val s &&. (report ^. GameReportVictory) ==. val v)
+    leagueFilter =
+      spec.leagues <&> \case
+        [] -> isNothing_ (report ^. GameReportLeague)
+        leagues -> toListFilter report GameReportLeague . map Just $ leagues
+    tokensFilter = toInequalityFilter report GameReportActionTokens <$> spec.tokens
+    dwarvenRings = toInequalityFilter report GameReportDwarvenRings <$> spec.dwarvenRings
+    corruption = toInequalityFilter report GameReportCorruption <$> spec.corruption
+    mordor = toMaybeInequalityFilter report GameReportMordor <$> spec.mordor
+    aragorn = toMaybeInequalityFilter report GameReportAragornTurn <$> spec.aragorn
+    treebeard =
+      spec.treebeard <&> \case
```

Comment:
Sort of opinionated on this one: I'm assuming a filter of `Treebeard = False` should give all games in which Treebeard was available but not mustered, AND all games that didn't use the Treebeard expansion at all. But willing to change this if you prefer different semantics.

Reply:
Okay yes that makes sense! Not valuable in that case even if it were easy, I wasn't thinking realistically about the data